### PR TITLE
Refactors cache backends into plugins + memcached datagrid support

### PR DESCRIPTION
**Pull request by acbramley**
- Removes need for OpenshiftConfig entities, these should be removed after this is deployed
- Moves connection details into shp_orchestration.settings, this allows us to easily DI the Openshift Client class for easier automated testing. OrchestrationProvider plugins are probably redundant now as well and could be removed.
- Moves Redis functionality into CacheBackend plugin
- Adds Memcached CacheBackend plugin with full test coverage!
- Adds cache backend field to Environments
- Fires cache backend functionality via getEnvironmentVariables, onEnvironmentCreate, and onEnvironmentDelete
- Adds sample datagrid openshift config

Relies on https://github.com/universityofadelaide/openshift-client/pull/39

### CODE REVIEW

File: web/modules/custom/shepherd/shp_orchestration/src/Plugin/OrchestrationProvider/OpenShiftOrchestrationProvider.php [hunk 1]
```diff
@@ -283,40 +283,20 @@ public function createdEnvironment(
...
-              $cron_jobs,
-              $image,
-              $cron_volumes,
-              $deploy_data
-            );
-          }
-        }
-      }
+    $image = $this->getImageStreamImage($sanitised_project_name, $sanitised_source_ref);
+    if ($image) {
```
Comment (acbramley) [inline web/modules/custom/shepherd/shp_orchestration/src/Plugin/OrchestrationProvider/OpenShiftOrchestrationProvider.php]: You can shorten this to `if ($image = ....`

Comment (pingers) [reply web/modules/custom/shepherd/shp_orchestration/src/Plugin/OrchestrationProvider/OpenShiftOrchestrationProvider.php]: Yeah... I got rid of the assignment in condition :smile:
I don't mind either way... it just reads a bit cleaner separating them.

**Comment by singularo** [comment]
Looks like the config entity is still in use - clicking Provider Settings here - /admin/config/shepherd/orchestration dies with:

```
web_1        | WATCHDOG: [NOTICE] [php] Notice: Undefined index: config_entity_id in Drupal\shp_orchestration\Controller\OrchestrationProviderSettingsController-&amp;gt;index() (line 76 of /code/web/modules/custom/shepherd/shp_orchestration/src/Controller/OrchestrationProviderSettingsController.php) #0 /code/web/core/includes/bootstrap.inc(587): _drupal_error_handler_real(8, &amp;#039;Undefined index...&amp;#039;, &amp;#039;/code/web/modul...&amp;#039;, 76, Array)
web_1        | #1 /code/web/modules/custom/shepherd/shp_orchestration/src/Controller/OrchestrationProviderSettingsController.php(76): _drupal_error_handler(8, &amp;#039;Undefined index...&amp;#039;, &amp;#039;/code/web/modul...&amp;#039;, 76, Array)
```

File: web/modules/custom/shepherd/shp_cache_backend/src/Plugin/CacheBackend/MemcachedDatagrid.php [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+   * Gets the memcache name for this environment.
+   *
+   * @param \Drupal\node\NodeInterface $environment
+   *   The environment name.
+   *
+   * @return string
+   *   The memcache name.
+   */
+  protected static function getMemcacheName(NodeInterface $environment) {
+    return 'memcached_node' . $environment->id();
```
Comment (singularo) [inline web/modules/custom/shepherd/shp_cache_backend/src/Plugin/CacheBackend/MemcachedDatagrid.php]: No '-' separator?

Comment (acbramley) [reply web/modules/custom/shepherd/shp_cache_backend/src/Plugin/CacheBackend/MemcachedDatagrid.php]: Was just following what's there at the moment on your cluster.

Comment (pingers) [reply web/modules/custom/shepherd/shp_cache_backend/src/Plugin/CacheBackend/MemcachedDatagrid.php:331]: I think this got changed - was just what RedHat had configured. We'd prefer the dash

File: openshift-config/datagrid/datagrid-app-ss.yml [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+              name: jolokia
+              protocol: TCP
+            - containerPort: 9779
+              name: prometheus
+              protocol: TCP
+            - containerPort: 8888
+              name: ping
+              protocol: TCP
+            - containerPort: 11211
+              name: mc-sheperd
```
[the remark this answers is not in the document]
Comment (acbramley) [reply openshift-config/datagrid/datagrid-app-ss.yml:64]: woops!
